File: src/backend/folha/src/model/folha_lancamento_comissao_model.py
from src import db
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel


class FolhaLancamentoComissaoModel(db.Model):
    __tablename__ = 'folha_lancamento_comissao'

    id = db.Column(db.Integer, primary_key=True)
    competencia = db.Column(db.String(7))
    vencimento = db.Column(db.DateTime)
    base_calculo = db.Column(db.Float)
    valor_comissao = db.Column(db.Float)
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_colaborador = data.get('idColaborador')
        self.competencia = data.get('competencia')
        self.vencimento = data.get('vencimento')
        self.base_calculo = data.get('baseCalculo')
        self.valor_comissao = data.get('valorComissao')

    def serialize(self):
        return {
            'id': self.id,
            'idColaborador': self.id_colaborador,
            'competencia': self.competencia,
            'vencimento': self.vencimento.isoformat(),
            'baseCalculo': self.base_calculo,
            'valorComissao': self.valor_comissao,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
        }

File: src/frontend/folha/lib/app/data/model/folha_lancamento_comissao_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:folha/app/infra/infra_imports.dart';
import 'package:folha/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class FolhaLancamentoComissaoModel {
	int? id;
	int? idColaborador;
	String? competencia;
	DateTime? vencimento;
	double? baseCalculo;
	double? valorComissao;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;

	FolhaLancamentoComissaoModel({
		this.id,
		this.idColaborador,
		this.competencia,
		this.vencimento,
		this.baseCalculo,
		this.valorComissao,
		this.viewPessoaColaboradorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'competencia',
		'vencimento',
		'base_calculo',
		'valor_comissao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Competencia',
		'Vencimento',
		'Base Calculo',
		'Valor Comissao',
	];

	FolhaLancamentoComissaoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idColaborador = jsonData['idColaborador'];
		competencia = jsonData['competencia'];
		vencimento = jsonData['vencimento'] != null ? DateTime.tryParse(jsonData['vencimento']) : null;
		baseCalculo = jsonData['baseCalculo']?.toDouble();
		valorComissao = jsonData['valorComissao']?.toDouble();
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['competencia'] = Util.removeMask(competencia);
		jsonData['vencimento'] = vencimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(vencimento!) : null;
		jsonData['baseCalculo'] = baseCalculo;
		jsonData['valorComissao'] = valorComissao;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		competencia = plutoRow.cells['competencia']?.value;
		vencimento = Util.stringToDate(plutoRow.cells['vencimento']?.value);
		baseCalculo = plutoRow.cells['baseCalculo']?.value?.toDouble();
		valorComissao = plutoRow.cells['valorComissao']?.value?.toDouble();
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
	}	

	FolhaLancamentoComissaoModel clone() {
		return FolhaLancamentoComissaoModel(
			id: id,
			idColaborador: idColaborador,
			competencia: competencia,
			vencimento: vencimento,
			baseCalculo: baseCalculo,
			valorComissao: valorComissao,
		);			
	}

	
}